### DISCUSSION TITLE
How to plot Inserts entities

### DISCUSSION THREAD

**Discussion by cristobalinho** [mozman/ezdxf#862] [Q&A]
Hello everybody, I've been trying to plot some INSERT entities using the transformation matrix and it works well when the block which is being referenced is only made of LINES entities, but when the block has LWPOLYLINE or ARC entities it doesn't work. I can solve this by converting the LWPOLYLINE into separate segments of LINES but Ihaven't done yet for ARC and I was asking if there is a more general way to do it for all entities. Here is part of my code.

```Python
import ezdxf

doc = ezdxf.readfile("File.dxf")
msp = doc.modelspace()

def Plot_insert(entity,ax):
    if entity.dxftype() == 'INSERT':
        block_name = entity.dxf.name
        block      = doc.blocks[block_name]
        entities   = block.query('*')
        # Define the transformation matrix
        matrix = entity.ucs().matrix                
        #matrix = entity.matrix44()   #Same matrix transformation      

        for e in entities:
            # Apply the scaling, rotation, and translation
            if e.dxftype()=='LINE':
                f = e.copy()
                f.transform(matrix) 
                # Now we plot the transformed entity
                Plot_entity(f,ax)
            elif e.dxftype() == 'LWPOLYLINE':
                epoints = list(e.get_points('xys'))               
                Newpoints = []
                for i in epoints:
                    Newpoints.append(entity.ucs().to_wcs(i))   
                ex = [p[0] for p in Newpoints]                 
                ey = [p[1] for p in Newpoints]            
                ax.plot(ex, ey, 'k')
```

I've also tried using other transformation matrix, and it happened that I could plot every entity but the coordinates weren't the right ones. Here is an example of what I get using the following matrix:

```Python
        x, y, z  = entity.dxf.insert
        sx,sy,sz = entity.dxf.xscale, entity.dxf.yscale, entity.dxf.zscale
        angle    = math.radians(entity.dxf.rotation)
        #angle    = entity.dxf.rotation
        # Define the transformation matrix
        #matrix = ezdxf.math.Matrix44.scale(sx, sy, sz).z_rotate(angle).translate(x, y, z)
```

The figure in the left is using the last matrix and the figure in the right is using the function Plot_insert, and give me the right position of the INSERT. Summarizing I need to know how to get a transformation matrix to plot any kind of entity in a block reference (INSERT).
I hope I was clear and thanks in advance!

![Figure 44](https://user-images.githubusercontent.com/130383135/230933917-5583dc98-fbd4-406e-9210-2871cc8befd1.png)

**Comment by mozman** [comment]
I have not read the full code, but it seems you want the render DXF content 😄.

You can learn from the `drawing` add-on how to do it. The INSERT entity is rendered by using the `Insert.virtual_entities()` method.

https://github.com/mozman/ezdxf/blob/be570c7be3d9b424052cefc5838c55a4623b4232/src/ezdxf/addons/drawing/frontend.py#L696

```Python
        def draw_insert(insert: Insert):
            self.draw_entities(insert.attribs)
            # draw_entities() includes the visibility check:
            self.draw_entities(
                insert.virtual_entities(
                    skipped_entity_callback=self.skip_entity
                    # TODO: redraw_order=True?
                )
            )
````

and the core `virtual_block_reference_entities()` function can be found in the `explode.py` module:

https://github.com/mozman/ezdxf/blob/be570c7be3d9b424052cefc5838c55a4623b4232/src/ezdxf/explode.py#L159

**Reply by cristobalinho** [reply]
Hello sir, thank you very much for your quick response! I'm going to look the drawing add-on to render the Inserts that I need to plot. Thanks again!

**Reply by cristobalinho** [reply]
Hello again! I was trying to use virtual_entities() method, but I have a problem plotting a LWPOLYLINE because I think it doesn't take in account the Y-axis reflection due to xscale = -1. I also tried with virtual_block_reference_entities() but the results are the same. Here I attach a image showing in the right side the correct ploting using Plot_insert and in the left side the results using virtual_entities() and the Plot_entity function I wrote.

```
def Plot_entity(entity,ax):
    if entity.dxftype() == 'POINT':
        x, y, _ = entity.dxf.location
        ax.plot(x, y,'k')
    elif entity.dxftype() == 'LINE':
        x1, y1, _ = entity.dxf.start
        x2, y2, _ = entity.dxf.end
        ax.plot([x1, x2], [y1, y2], 'k')
    elif entity.dxftype() == 'CIRCLE':
        x, y, _ = entity.dxf.center
        r = entity.dxf.radius
        circle = plt.Circle((x, y), r, fill=False)
        ax.add_artist(circle)
    elif entity.dxftype() == 'LWPOLYLINE':
        # Get the polyline points
        points = list(entity.get_points('xy'))
        # Create a list of x and y coordinates
        x = [p[0] for p in points]
        y = [p[1] for p in points]
        # Plot the polyline segments as a single line
        ax.plot(x, y, 'k')
    elif entity.dxftype() == 'ARC':
        Plot_ARC(entity,ax)
    else:
        print('Plotting',entity.dxftype(),'has not been defined yet')
```

![Figure 110](https://user-images.githubusercontent.com/130383135/231297626-96fdec9f-19ff-4100-855c-7fee4fe0f2b7.png)

I also tried to use the drawing add-on to plot some inserts, but I'm not ver knowledgeable with classes. I wrote

```
import sys
import matplotlib.pyplot as plt
from ezdxf import recover
from ezdxf.addons.drawing import RenderContext, Frontend, Properties
from ezdxf.addons.drawing.matplotlib import MatplotlibBackend

# Safe loading procedure (requires ezdxf v0.14):
try:
    doc, auditor = recover.readfile('AM-101-PISO 2.dxf')
except IOError:
    print(f'Not a DXF file or a generic I/O error.')
    sys.exit(1)
except ezdxf.DXFStructureError:
    print(f'Invalid or corrupted DXF file.')
    sys.exit(2)

# The auditor.errors attribute stores severe errors,
# which may raise exceptions when rendering.
msp = doc.modelspace()
Inserts = msp.query('INSERT')
if not auditor.has_errors:
    fig = plt.figure()
    ax = fig.add_axes([0, 0, 1, 1])
    ctx = RenderContext(doc)
    out = MatplotlibBackend(ax)


Frontend(ctx, out).draw_composite_entity(Inserts[0],Properties())
```

but I get an empty figure. I would want to know how to use virtual_entities() properly I hope don't disturb you
Thanks!
